File: openmole-ui-ide/openmole-ide-ant/org.openmole.ui.ide/src/org/openmole/ui/ide/workflow/implementation/CapsuleModelUI.java
/*
 *  Copyright (C) 2010 leclaire
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the Affero GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.ui.ide.workflow.implementation;

import java.util.HashSet;
import java.util.Set;
import org.openmole.ui.ide.workflow.model.ICapsuleModelUI;
import org.openmole.core.model.capsule.IGenericCapsule;
import org.openmole.core.model.task.IGenericTask;
import org.openmole.ui.ide.commons.ApplicationCustomize;
import org.openmole.ui.ide.workflow.model.IGenericTaskModelUI;

/**
 *
 * @author Mathieu Leclaire <mathieu.leclaire@openmole.fr>
 */
public class CapsuleModelUI<T extends IGenericCapsule> extends ObjectModelUI implements ICapsuleModelUI {

    public static CapsuleModelUI EMPTY_CAPSULE_MODEL = new CapsuleModelUI();
    private IGenericTaskModelUI<IGenericTask> taskModel;
    private transient int nbInputSlots = 0;
    private boolean startingCapsule = false;
    private final static String category = "Task Tapsules";
    private Set<ICapsuleModelUI> connectedTo = new HashSet<ICapsuleModelUI>();
    private boolean containsTask = false;

    CapsuleModelUI() {
        this(TaskModelUI.EMPTY_TASK_MODEL);
    }

    CapsuleModelUI(IGenericTaskModelUI<IGenericTask> taskModel) {
        this.taskModel = taskModel;
    }

    public boolean containsTask() {
        return containsTask;
    }

    @Override
    public IGenericTaskModelUI<IGenericTask> getTaskModel() {
        return taskModel;
    }

    @Override
    public void setTaskModel(IGenericTaskModelUI taskModel) {
        this.taskModel = taskModel;
        this.containsTask = true;
    }

    @Override
    public String getCategory() {
        return category;
    }

    @Override
    public int getNbInputslots() {
        return nbInputSlots;
    }

    @Override
    public void addInputSlot() {
        nbInputSlots++;
    }

    @Override
    public boolean isSlotRemovable() {
        return (nbInputSlots > 1 ? true : false);
    }

    @Override
    public boolean isSlotAddable() {
        return (nbInputSlots < ApplicationCustomize.NB_MAX_SLOTS ? true : false);
    }

    @Override
    public void removeInputSlot() {
        nbInputSlots -= 1;
    }

    @Override
    public void eventOccured(Object t) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void addTransition(ICapsuleModelUI taskmodel) {
        connectedTo.add(taskmodel);
    }

    @Override
    public void defineAsStartingCapsule() {
        nbInputSlots = 1;
        startingCapsule = true;
    }

    @Override
    public void defineAsRegularCapsule() {
        startingCapsule = false;
    }

    @Override
    public boolean isStartingCapsule() {
        return startingCapsule;
    }

    @Override
    public boolean hasChild() {
        return !connectedTo.isEmpty();
    }

    @Override
    public Set getChilds() {
        return connectedTo;
    }
}

File: platform/core/src/main/scala/org/openmole/ide/core/workflow/implementation/CapsuleModelUI.scala
/*
 * Copyright (C) 2011 Mathieu leclaire <mathieu.leclaire at openmole.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.ide.core.workflow.implementation

import org.openmole.ide.core.commons.CapsuleType._
import org.openmole.ide.core.workflow.model.ICapsuleModelUI
import org.openmole.ide.core.palette.ElementFactories
import org.openmole.ide.core.palette.PaletteElementFactory

class CapsuleModelUI(var dataProxy: Option[PaletteElementFactory] = None, var nbInputSlots: Int = 0) extends ICapsuleModelUI{

  val category= "Task Tapsules"
  var startingCapsule = false
 // var containsTask = false
 var capsuleType = CAPSULE
  
  //def this(pef: PaletteElementFactory)= this(Some(pef))
 
  def addInputSlot= nbInputSlots+= 1
  
  def removeInputSlot= nbInputSlots-= 1
  
  def setDataProxy(pef: PaletteElementFactory)={
    dataProxy= Some(pef)
    if (ElementFactories.isExplorationTaskFactory(pef.panelUIData)) capsuleType = EXPLORATION_TASK else capsuleType = BASIC_TASK
   // containsTask= true
  }
  
  override def defineStartingCapsule(on: Boolean)= {
    startingCapsule= on
    if (on) nbInputSlots= 1
  }
}